aten/src/ATen/native/xnnpack/OpContext.h

```diff
@@ -2,95 +2,131 @@
 
 #ifdef USE_XNNPACK
 
+#include <ATen/Tensor.h>
 #include <ATen/core/ivalue.h>
 #include <ATen/native/xnnpack/Common.h>
-#include <ATen/Tensor.h>
 
 namespace at {
 namespace native {
 namespace xnnpack {
 
-using SerializationTypeLinearPrePack = std::tuple<Tensor, c10::optional<Tensor>>;
-using SerializationTypeConv2dPrePack =
-  std::tuple<Tensor, c10::optional<Tensor>,
-  std::vector<int64_t>, std::vector<int64_t>, std::vector<int64_t>, int64_t>;
-
-class XNNPackLinearOpContext : public torch::jit::CustomClassHolder {
-  private:
-    Tensor orig_weight_;
-    c10::optional<Tensor> orig_bias_;
-    ContextLinear op_context_;
-
-  public:
-    XNNPackLinearOpContext(Tensor&& weight,
-        c10::optional<Tensor>&& bias,
-        ContextLinear&& op_context) :
-      orig_weight_(std::move(weight)),
-      orig_bias_(std::move(bias)),
-      op_context_(std::move(op_context)) {}
-
-    const ContextLinear& get_context() const {
-      return op_context_;
-    }
-
-    SerializationTypeLinearPrePack unpack() {
-      return std::make_tuple(orig_weight_, orig_bias_);
-    }
-
-    static c10::intrusive_ptr<XNNPackLinearOpContext> create_context(Tensor&& weight,
-        c10::optional<Tensor>&& bias,
-        const c10::optional<double> output_min,
-        const c10::optional<double> output_max);
+using SerializationTypeLinearPrePack = std::tuple<
+    Tensor,
+    c10::optional<Tensor>,
+    c10::optional<double>,
+    c10::optional<double>>;
+using SerializationTypeConv2dPrePack = std::tuple<
+    Tensor,
+    c10::optional<Tensor>,
+    std::vector<int64_t>,
+    std::vector<int64_t>,
+    std::vector<int64_t>,
+    int64_t,
+    c10::optional<double>,
+    c10::optional<double>>;
+
+class LinearOpContext : public torch::jit::CustomClassHolder {
+ protected:
+  Tensor orig_weight_;
```

Comment:
this duplicates all weights for serialization, I guess xnnpack doesn't have "unpack" method, does it?

Reply:
This is correct. My plan was to have the weights set to null inside __setstate__ conditioned on some flags such as lite_interpreter run etc., that ensure we will not be serializing this. Is that the right approach?

Reply:
@dzhulgakov, so I think we dont need to use trampoline. However, given that we are doing all this to make a generic API, it makes sense to hide backend specific context creation behind something, which in this PR turns out to be just a trampoline. However I am gonna stack a PR on top of this which will check global context to see which engine we want to use for packing (for now xnnpack) and direct context creation appropriately. I will refactor this code so that both __setstate__ method and PrePack op are directed to a common method for context creaton.

Reply:
I'm not sure what the right approach is here.  If we set the weights to null, that means we can't re-save the model, right?

Reply:
Yes, but if we know we will not need to, e.g. when we know we are doing inference in lite interpreter, it should ok, right?

Reply:
Yeah, but what about people using the full JIT.  This is okay for now, but we should keep thinking about better solutions.

Reply:
Keeping the weights around is fine, we can gate freeing them with mobile interpreter flag indeed (we kind of do that for qnnpack on mobile I think).